alf/experience_replayers/replay_buffer.py

```diff
@@ -466,6 +517,51 @@ def _pad(self, x, env_ids):
         return ((self._current_pos[env_ids] - x - 1) //
                 self._max_length) * self._max_length + x
 
+    def _set_default_return(self, env_ids):
+        mask = torch.ones_like(env_ids, dtype=torch.bool)
+        if torch.any(mask):
+            current_pos = self._current_pos[env_ids]
+            current_pos -= 1
+            ind = (env_ids[mask], self.circular(current_pos[mask]))
+            self._episodic_discounted_return[ind] = self._default_return
+
+    def _compute_store_episodic_return(self, env_ids, discount):
+        # Always pass in env_ids whose step_type is LAST, to save computation.
+        # overall mask
+        mask = torch.ones_like(env_ids, dtype=torch.bool)
+        # mask that shrinks as we are done with some env_ids
+        discounted_return = torch.zeros_like(env_ids, dtype=torch.float32)
+        if torch.any(mask):
+            mask &= discount == 0
+            env_ids = env_ids[mask]
+            current_pos = self._current_pos[env_ids]
+            current_pos -= 1
+            # Only store episodic return when game ends, not ending with timelimit.
+        while torch.any(mask):
```

Comment:
When num_parallel_environments is large, this can cost a lot of time since there will be lots of game endings.
It is possible to do this without using loop, similar to what MCTSAlgorithm._backup() does.